tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/gatt_client_test.c

```diff
@@ -0,0 +1,363 @@
+				.uuid = BT_UUID_GATT_DB_HASH,
+			},
+		},
+		.expected_status = BT_ATT_ERR_SUCCESS,
+	},
+	{
+		 /* Read on channel 1 */
+		.params = {
+			.func = gatt_read_cb,
+			.handle_count = 1,
+			.single = {
+				.handle = 0, /* Will be set later */
+				.offset = 0,
+			},
+		},
+		.expected_status = BT_ATT_ERR_SUCCESS,
+	},
+};
+
+static volatile uint8_t read_index;
+
+static uint8_t gatt_read_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_read_params *params,
+			    const void *data, uint16_t length)
+{
+	printk("GATT read cb: err 0x%02X\n", err);
+	if (err != reads[read_index].expected_status) {
+		FAIL("Unexpected status from read: 0x%02X\n", err);
+	}
+
+	if (params != &reads[read_index].params) {
+		FAIL("Wrong read\n");
+	}
+
+	(void)memset(params, 0, sizeof(*params));
+
+	read_index++;
+
+	return 0;
+}
+
+static void gatt_read(struct bt_gatt_read_params *read_params)
+{
+	int err;
+
+	printk("Reading\n");
+
+	err = bt_gatt_read(g_conn, read_params);
+	if (err != 0) {
+		FAIL("bt_gatt_read failed: %d\n", err);
+	}
+}
+
+static void write_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_write_params *params)
+{
+	if (err != BT_ATT_ERR_SUCCESS) {
+		FAIL("Write failed: 0x%02X\n", err);
+	}
+
+	SET_FLAG(flag_write_complete);
+}
+
+static void enable_robust_caching(void)
+{
+	/* Client Supported Features Characteristic Value
+	 * Bit 0: Robust Caching
+	 * Bit 1: EATT
+	 */
+	static const uint8_t csf[] = { BIT(0) | BIT(1) };
+	static struct bt_gatt_write_params write_params = {
+		.func = write_cb,
+		.offset = 0,
+		.data = csf,
+		.length = sizeof(csf),
+	};
+	int err;
+
+	printk("Writing to Client Supported Features Characteristic\n");
+
+	write_params.handle = csf_handle;
+	UNSET_FLAG(flag_write_complete);
+
+	err = bt_gatt_write(g_conn, &write_params);
+	if (err) {
+		FAIL("bt_gatt_write failed (err %d)\n", err);
+	}
+
+	WAIT_FOR_FLAG(flag_write_complete);
+	printk("Success\n");
+}
+
+static void test_main(void)
+{
+	int err;
+
+	err = bt_enable(NULL);
+	if (err != 0) {
+		FAIL("Bluetooth discover failed (err %d)\n", err);
+	}
+
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, device_found);
+	if (err != 0) {
+		FAIL("Scanning failed to start (err %d)\n", err);
+	}
+
+	printk("Scanning successfully started\n");
+
+	WAIT_FOR_FLAG(flag_is_connected);
+
+	gatt_discover(test_svc_uuid, BT_GATT_DISCOVER_PRIMARY);
+	gatt_discover(BT_UUID_GATT_CLIENT_FEATURES, BT_GATT_DISCOVER_CHARACTERISTIC);
+
+	enable_robust_caching();
+
+	while (bt_eatt_count(g_conn) < 1) {
+		/* Wait for EATT channel to connect, in case it hasn't already */
+		k_sleep(K_MSEC(10));
+	}
+
+	k_sleep(K_MSEC(1000)); /* Wait for new service to be added by server */
+
+	reads[0].params.single.handle = chrc_handle;
+	reads[1].params.single.handle = chrc_handle;
+	reads[3].params.single.handle = chrc_handle;
+
+	gatt_read(&reads[0].params); /* Read characteristic, channel 1 */
+	gatt_read(&reads[1].params); /* Read characteristic, channel 2 */
```

Comment:
I don't think there is a way currently for the application to see which channel a request is received on. We only know that these are sent on different channel because they are queued right after each other, before a response is received. When we have some API for sending on EATT channels this can be ensured by the client.